[WIP] Terraform Docs for TF policy (#2690)
hashicorp/web-unified-docs: content/terraform-policy/v0.1.x (beta)/docs/policy/compare.mdx

---
page_title: Compare policy frameworks
description: >-
  Learn about how Terraform policy compares to other policy frameworks.
---

# Compare policy frameworks

> Terraform policy is a declarative, HCL-based policy as code solution deeply integrated with Terraform. It provides native support for evaluating policies at multiple stages of the Terraform lifecycle, from initialization through post-apply validation.

Reviewer: ```suggestion
Terraform policy is a declarative, HCL-based policy as code solution deeply integrated with Terraform. It provides native support for evaluating policies at multiple stages of the Terraform lifecycle, from initialization through post-apply validation. We recommend using Terraform policy to enforce governance in Terraform workflows.
```

@include "partials/beta.mdx"

## Comparison with Sentinel

> Terraform policy is the policy as code framework HashiCorp recommends for Terraform workflows.

Reviewer: ```suggestion
Sentinel is a purpose-built policy as code framework. Refer the [Sentinel documentation](/sentinel) for more information. 
```

| Feature | Sentinel | Terraform policy |
|---------|----------|------------------|

> | Policy language | Sentinel DSL - a purpose-built policy language with its own syntax and type system | HCL: the same configuration language used to write Terraform |

Reviewer: ```suggestion
| Policy language | Sentinel DSL, which is a purpose-built policy language with a unique syntax and type system | HCL, which is the same  language for writing Terraform configuration |
```

> | Enforcement levels | Advisory, soft-mandatory, hard-mandatory | Advisory, Mandatory-overridable, Mandatory |

Reviewer 2: ```suggestion
| Enforcement levels | Advisory, soft-mandatory, hard-mandatory | Advisory, Mandatory_overridable, Mandatory |
```

> | Enforcement levels | Advisory, soft-mandatory, hard-mandatory | Advisory, Mandatory-overridable, Mandatory |

Reviewer: ```suggestion
| Enforcement levels | Advisory, soft-mandatory, hard-mandatory | Advisory, mandatory-overridable, mandatory |
```
I don't know how meaningful this comparison is. You might be able to infer what "advisory" and "mandatory" mean, but without explanation, how is anyone going to know the difference between "soft-mandatory" and "mandatory-overridable" ? Consider removing this row or replacing with links to the explanations.

> | Resource relationships | Requires manually iterating resource_changes and comparing each attribute | Provides a functional approach to query related resources |

Reviewer: ```suggestion
| Resource relationship discovery | Manually iterate on resource changes and compare each attribute | Provides a functional approach to query related resources |
```
Right? We might need to include more detail because if I didn't already read all the docs I probably wouldn't have any idea what this means.

| Data source lookups | Cannot trigger or query data sources at evaluation time | Provides a functional approach to query data source results during policy evaluation |

> | Init phase evaluation | Runs after `terraform init` is complete: providers and modules are already downloaded | Runs during `terraform init` when possible, before Terraform downloads providers and modules |

Reviewer: ```suggestion
| Initialization phase evaluation | Runs after `terraform init` completes and providers and modules are already downloaded | When possible, runs during `terraform init` and before Terraform downloads providers and modules |
```

| Plan phase evaluation | Runs after the plan phase is complete | Runs after the plan phase is complete |

> | Apply phase evaluation | Only runs during plan/policy check stage | Runs after the apply phase completes to validate values that are only known after apply |

Reviewer: ```suggestion
| Apply phase evaluation | Only runs during plan and policy check stage | Runs after the apply phase completes to validate values that are only known after apply |
```

## Comparison with OPA

> Open Policy Agent (OPA) is a general-purpose policy engine that can be integrated with HCP Terraform. While OPA is powerful for cross-platform policy enforcement, Terraform policy provides deeper integration with Terraform workflows.

Reviewer: ```suggestion
Open Policy Agent (OPA) is a general-purpose policy engine that you can integrate with HCP Terraform. While OPA is powerful for cross-platform policy enforcement, Terraform policy provides deeper integration with Terraform workflows.
```

| Feature | OPA (HCP Terraform) | Terraform policy |
|---------|---------------------|------------------|

> | Policy language | Rego - a general-purpose declarative query language native to the OPA ecosystem | HCL - the same configuration language used to write Terraform |

Reviewer: ```suggestion
| Policy language | Rego, which is a general-purpose declarative query language native to the OPA ecosystem | HCL, which is the same language for writing Terraform configuration |
```

> | Enforcement levels | Pass/fail only - no built-in concept of advisory or soft-mandatory levels | Advisory, Mandatory-overridable, Mandatory |

Reviewer 2: ```suggestion
| Enforcement levels | Pass/fail only - no built-in concept of advisory or soft-mandatory levels | Advisory, Mandatory_overridable, Mandatory |
```

> | Enforcement levels | Pass/fail only - no built-in concept of advisory or soft-mandatory levels | Advisory, Mandatory-overridable, Mandatory |

Reviewer: ```suggestion
| Enforcement levels | Pass or fail. No built-in concept of advisory or soft-mandatory levels | Advisory, mandatory-overridable, mandatory |
```
Same question as line 20

> | Resource relationships | Requires parsing the full plan JSON: no structured helper API | Provides a functional approach to query related resources |

Reviewer: ```suggestion
| Resource relationship discovery | You must parse the full plan. No structured helper API | Provides a functional approach to query related resources |
```

| Data source lookups | Cannot trigger or query data sources at evaluation time | Provides a functional approach to query data source results during policy evaluation |

> | Init phase evaluation | Runs after `terraform init` is complete: providers and modules are already downloaded | Runs during `terraform init` when possible, before Terraform downloads providers and modules |

Reviewer: ```suggestion
| Initialization phase evaluation | Runs after `terraform init` completes and providers and modules are already downloaded | When possible, runs during `terraform init` before Terraform downloads providers and modules |
```

| Plan phase evaluation | Runs after the plan phase is complete | Runs after the plan phase is complete |

> | Apply phase evaluation | Only runs during plan/policy check stage | Runs after the apply phase completes to validate values that are only known after apply |

Reviewer: ```suggestion
| Apply phase evaluation | Only runs during plan and policy check stage | Runs after the apply phase completes to validate values that are only known after apply |
```

- **Early validation**: Provider and module policies are evaluated during `terraform init`, before any downloads occur.
- **Post-apply evaluation**: Policies can validate provider-computed values (ARNs, IDs, ports) that are only known after resources are provisioned.
- **Resource relationship queries**: Built-in functions like `core::getresources()` make it easy to validate relationships between resources.
- **Data source lookups**: Policies can query data source results directly using `core::getdatasource()` to enforce dynamic requirements.

### Flexible enforcement levels

Terraform policy supports three enforcement levels:

- **Advisory**: HCP Terraform will warn the user of the policy violation, but will not block the run.
- **Mandatory-overridable**: HCP Terraform will block the run if the policy fails before the apply phase, but a user with the `Manage policy overrides` permission can override the policy violation and continue the run.
- **Mandatory**: HCP Terraform will blocks the run if the policy fails before the apply phase.

### Support for Terraform workflows

Terraform policy works Terraform workspaces, covering multi-component and multi-environment deployments.

> ## Comparison with Sentinel

Reviewer: Per the comment on line 2 of this file, I think these tables are good content for an "Overview" section

#### OPA

In OPA, because the full plan JSON is available, you can cross-reference resource types in Rego, but there is no structured helper API. Comparing two resource types by a shared attribute requires raw array comprehensions over `input.plan.resource_changes`. Deeply nested relationships across multiple resource types requires additional effort to write and maintain.

### Evaluate policies using data source lookups

Use cases:

- Enforce that EC2 instances use only AMIs from an approved `aws_ami` data source lookup
- Validate that an RDS instance uses a KMS key that exists in an `aws_kms_key` lookup
- Check that an S3 bucket region matches a known approved region list retrieved via a data source

#### Terraform policy

> Terraform policy provides the `core::getdatasource(type, name)` function, which queries data source results directly at policy evaluation time. For example, you can look up approved AMI IDs from an `aws_ami` data source and enforce that EC2 instances use only those AMIs.

Reviewer 3: `core::getdatasource`'s second argument is a filter/attributes map, not a name (matches the getdatasource reference page). Worth updating this to `core::getdatasource(type, filter)`.
